In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
%matplotlib inline

In [2]:
import os #The OS module in Python provides functions for creating and removing a directory 
#(folder), fetching its contents, changing and identifying the current directory
import glob #glob (short for global) is used to return all file paths that match a specific pattern.

In [3]:
#function to concat all 12 months csv files 
def get_merged_df(data_folder_path: str) -> pd.DataFrame:
    files = list(os.listdir(data_folder_path)) #list of csv files in that directory
    file_paths = [data_folder_path + f for f in files] #location for each csv file 
    #load all files
    #df_per_file is a generator object where all the csv are read 
    df_per_file = (pd.read_csv(fp, 
                               low_memory = False, #data chunk (packet data with mixed type inference)
                               parse_dates = ['started_at', 'ended_at'], #converted to datetime datatype
                               infer_datetime_format = True,
                               memory_map = True)#map the file object directly onto memory and access the data directly from there
                  for fp in file_paths)
    return pd.concat(df_per_file, ignore_index = True) # all files are concatenated 
# calling the function passing the location of DATA folder
df = get_merged_df(os.getcwd() + "\\DATA\\") 

In [4]:
os.getcwd() + "\\DATA\\"    #gives the location 

'C:\\Users\\kishore kumar\\Desktop\\google data\\CASE STUDY PROJECT\\CASE STUDY 1 CYCLIST PROJECT\\DATA\\'

In [5]:
files = list(os.listdir(os.getcwd() + "\\DATA\\"))
files #all the files inside the folder Data

['202104-divvy-tripdata.csv',
 '202105-divvy-tripdata.csv',
 '202106-divvy-tripdata.csv',
 '202107-divvy-tripdata.csv',
 '202108-divvy-tripdata.csv',
 '202109-divvy-tripdata.csv',
 '202110-divvy-tripdata.csv',
 '202111-divvy-tripdata.csv',
 '202112-divvy-tripdata.csv',
 '202201-divvy-tripdata.csv',
 '202202-divvy-tripdata.csv',
 '202203-divvy-tripdata.csv']

In [6]:
file_paths = [os.getcwd() + "\\DATA\\"  + f for f in files]
file_paths  #getting each file with their location

['C:\\Users\\kishore kumar\\Desktop\\google data\\CASE STUDY PROJECT\\CASE STUDY 1 CYCLIST PROJECT\\DATA\\202104-divvy-tripdata.csv',
 'C:\\Users\\kishore kumar\\Desktop\\google data\\CASE STUDY PROJECT\\CASE STUDY 1 CYCLIST PROJECT\\DATA\\202105-divvy-tripdata.csv',
 'C:\\Users\\kishore kumar\\Desktop\\google data\\CASE STUDY PROJECT\\CASE STUDY 1 CYCLIST PROJECT\\DATA\\202106-divvy-tripdata.csv',
 'C:\\Users\\kishore kumar\\Desktop\\google data\\CASE STUDY PROJECT\\CASE STUDY 1 CYCLIST PROJECT\\DATA\\202107-divvy-tripdata.csv',
 'C:\\Users\\kishore kumar\\Desktop\\google data\\CASE STUDY PROJECT\\CASE STUDY 1 CYCLIST PROJECT\\DATA\\202108-divvy-tripdata.csv',
 'C:\\Users\\kishore kumar\\Desktop\\google data\\CASE STUDY PROJECT\\CASE STUDY 1 CYCLIST PROJECT\\DATA\\202109-divvy-tripdata.csv',
 'C:\\Users\\kishore kumar\\Desktop\\google data\\CASE STUDY PROJECT\\CASE STUDY 1 CYCLIST PROJECT\\DATA\\202110-divvy-tripdata.csv',
 'C:\\Users\\kishore kumar\\Desktop\\google data\\CASE STUDY P

In [7]:
#creating a generator object and reading all the files in file_paths
df_per_file = (pd.read_csv(fp, 
                               low_memory = False, 
                               parse_dates = ['started_at', 'ended_at'], #cnverted to datetime datatype
                               infer_datetime_format = True,
                               memory_map = True)
                  for fp in file_paths)
for i in df_per_file:  #printing the generator object
    print(i)

                 ride_id  rideable_type          started_at  \
0       6C992BD37A98A63F   classic_bike 2021-04-12 18:25:36   
1       1E0145613A209000    docked_bike 2021-04-27 17:27:11   
2       E498E15508A80BAD    docked_bike 2021-04-03 12:42:45   
3       1887262AD101C604   classic_bike 2021-04-17 09:17:42   
4       C123548CAB2A32A5    docked_bike 2021-04-03 12:42:25   
...                  ...            ...                 ...   
337225  461A6B0728E063DB   classic_bike 2021-04-09 17:09:03   
337226  CF1D3A35E3654F6F    docked_bike 2021-04-04 13:27:08   
337227  4308ADB9171ACE49   classic_bike 2021-04-30 18:15:40   
337228  04DFB53077A17DA2  electric_bike 2021-04-18 11:40:37   
337229  DB6F78ABBECA378C   classic_bike 2021-04-23 19:22:16   

                  ended_at                   start_station_name  \
0      2021-04-12 18:56:55                State St & Pearson St   
1      2021-04-27 18:31:29             Dorchester Ave & 49th St   
2      2021-04-07 11:40:24                

                 ride_id  rideable_type          started_at  \
0       0A1B623926EF4E16    docked_bike 2021-07-02 14:44:36   
1       B2D5583A5A5E76EE   classic_bike 2021-07-07 16:57:42   
2       6F264597DDBF427A   classic_bike 2021-07-25 11:30:55   
3       379B58EAB20E8AA5   classic_bike 2021-07-08 22:08:30   
4       6615C1E4EB08E8FB  electric_bike 2021-07-28 16:08:06   
...                  ...            ...                 ...   
822405  1E660BF8DCDAAAB6  electric_bike 2021-07-04 10:40:41   
822406  A2448BDFD9B3653D  electric_bike 2021-07-04 12:47:41   
822407  2D612BF8530379DA  electric_bike 2021-07-03 21:41:58   
822408  6D615D18B765C9B3  electric_bike 2021-07-03 22:10:31   
822409  0F31D311323F0048  electric_bike 2021-07-04 07:03:50   

                  ended_at            start_station_name start_station_id  \
0      2021-07-02 15:19:58  Michigan Ave & Washington St            13001   
1      2021-07-07 17:16:09    California Ave & Cortez St            17660   
2      2021-

                 ride_id  rideable_type          started_at  \
0       620BC6107255BF4C  electric_bike 2021-10-22 12:46:42   
1       4471C70731AB2E45  electric_bike 2021-10-21 09:12:37   
2       26CA69D43D15EE14  electric_bike 2021-10-16 16:28:39   
3       362947F0437E1514  electric_bike 2021-10-16 16:17:48   
4       BB731DE2F2EC51C5  electric_bike 2021-10-20 23:17:54   
...                  ...            ...                 ...   
631221  BA077FDD42DABF12   classic_bike 2021-10-14 21:45:05   
631222  B7D99254E798AA6B   classic_bike 2021-10-02 15:28:28   
631223  BCCFD66DA466400F  electric_bike 2021-10-08 16:47:10   
631224  623E0F6F50CDD6E4   classic_bike 2021-10-08 07:49:47   
631225  83FA6AC52B7B7A7B   classic_bike 2021-10-02 12:55:45   

                  ended_at           start_station_name start_station_id  \
0      2021-10-22 12:49:50     Kingsbury St & Kinzie St     KA1503000043   
1      2021-10-21 09:14:14                          NaN              NaN   
2      2021-10-

                 ride_id  rideable_type          started_at  \
0       C2F7DD78E82EC875  electric_bike 2022-01-13 11:59:47   
1       A6CF8980A652D272  electric_bike 2022-01-10 08:41:56   
2       BD0F91DFF741C66D   classic_bike 2022-01-25 04:53:40   
3       CBB80ED419105406   classic_bike 2022-01-04 00:18:04   
4       DDC963BFDDA51EEA   classic_bike 2022-01-20 01:31:10   
...                  ...            ...                 ...   
103765  8788DA3EDE8FD8AB  electric_bike 2022-01-18 12:36:48   
103766  C6C3B64FDC827D8C  electric_bike 2022-01-27 11:00:06   
103767  CA281AE7D8B06F5A  electric_bike 2022-01-10 16:14:51   
103768  44E348991862319B  electric_bike 2022-01-19 13:22:11   
103769  E477C594A182AE58  electric_bike 2022-01-13 17:24:43   

                  ended_at             start_station_name start_station_id  \
0      2022-01-13 12:02:44       Glenwood Ave & Touhy Ave              525   
1      2022-01-10 08:46:17       Glenwood Ave & Touhy Ave              525   
2      20

In [8]:
df.head()

ride_id rideable_type          started_at            ended_at  \
0  6C992BD37A98A63F  classic_bike 2021-04-12 18:25:36 2021-04-12 18:56:55   
1  1E0145613A209000   docked_bike 2021-04-27 17:27:11 2021-04-27 18:31:29   
2  E498E15508A80BAD   docked_bike 2021-04-03 12:42:45 2021-04-07 11:40:24   
3  1887262AD101C604  classic_bike 2021-04-17 09:17:42 2021-04-17 09:42:48   
4  C123548CAB2A32A5   docked_bike 2021-04-03 12:42:25 2021-04-03 14:13:42   

         start_station_name start_station_id              end_station_name  \
0     State St & Pearson St     TA1307000061  Southport Ave & Waveland Ave   
1  Dorchester Ave & 49th St     KA1503000069      Dorchester Ave & 49th St   
2     Loomis Blvd & 84th St            20121         Loomis Blvd & 84th St   
3   Honore St & Division St     TA1305000034  Southport Ave & Waveland Ave   
4     Loomis Blvd & 84th St            20121         Loomis Blvd & 84th St   

  end_station_id  start_lat  start_lng    end_lat    end_lng member_casual  
0          13235  41.897448 -87.628722  41.948150 -87.663940        member  
1   KA1503000069  41.805772 -87.592464  41.805772 -87.592464        casual  
2          20121  41.741487 -87.658410  41.741487 -87.658410        casual  
3          13235  41.903119 -87.673935  41.948150 -87.663940        member  
4          20121  41.741487 -87.658410  41.741487 -87.658410        casual

In [9]:
df.tail()

ride_id  rideable_type          started_at  \
5723527  E23BE3DB740BE3B4    docked_bike 2022-03-13 16:31:03   
5723528  15AF71F3220F2A75    docked_bike 2022-03-09 06:56:02   
5723529  9C4CE6CC19F8225B  electric_bike 2022-03-09 15:55:26   
5723530  F4E136DEF696F3AE  electric_bike 2022-03-21 16:12:44   
5723531  5AEC5F39AF711895   classic_bike 2022-03-03 18:13:40   

                   ended_at      start_station_name start_station_id  \
5723527 2022-03-13 16:39:32   Michigan Ave & Oak St            13042   
5723528 2022-03-09 07:42:14    Broadway & Barry Ave            13137   
5723529 2022-03-09 16:08:54                     NaN              NaN   
5723530 2022-03-21 16:18:24                     NaN              NaN   
5723531 2022-03-03 18:23:39  Clark St & Randolph St     TA1305000030   

                     end_station_name end_station_id  start_lat  start_lng  \
5723527       Streeter Dr & Grand Ave          13022  41.900960 -87.623777   
5723528  Ravenswood Ave & Berteau Ave   TA1309000018  41.937582 -87.644098   
5723529     Albany Ave & Montrose Ave          15621  41.940000 -87.710000   
5723530     Larrabee St & Division St   KA1504000079  41.910000 -87.650000   
5723531       Streeter Dr & Grand Ave          13022  41.884576 -87.631890   

           end_lat    end_lng member_casual  
5723527  41.892278 -87.612043        casual  
5723528  41.957921 -87.673567        casual  
5723529  41.961041 -87.705866        member  
5723530  41.903486 -87.643353        member  
5723531  41.892278 -87.612043        member

In [10]:
df.shape #dataframe contains (rows ,columns)

(5723532, 13)

In [11]:
df.info()
#for knowing datatype of the columns

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5723532 entries, 0 to 5723531
Data columns (total 13 columns):
 #   Column              Dtype         
---  ------              -----         
 0   ride_id             object        
 1   rideable_type       object        
 2   started_at          datetime64[ns]
 3   ended_at            datetime64[ns]
 4   start_station_name  object        
 5   start_station_id    object        
 6   end_station_name    object        
 7   end_station_id      object        
 8   start_lat           float64       
 9   start_lng           float64       
 10  end_lat             float64       
 11  end_lng             float64       
 12  member_casual       object        
dtypes: datetime64[ns](2), float64(4), object(7)
memory usage: 567.7+ MB


In [12]:
missing=df.isnull().sum()
missing # finding the number of null values in each column 

ride_id                    0
rideable_type              0
started_at                 0
ended_at                   0
start_station_name    745376
start_station_id      745373
end_station_name      796247
end_station_id        796247
start_lat                  0
start_lng                  0
end_lat                 4716
end_lng                 4716
member_casual              0
dtype: int64

In [13]:
df["end_station_name"].isnull().sum() 
#finding no of null values in end_station_name column because it has max null values

796247

In [14]:
df.shape[0]  #total no of rows in the data

5723532

In [15]:
(df["end_station_name"].isnull().sum()/df.shape[0])*100
#13.9% of values is missing from a column at max 

13.911811797330738

In [16]:
#finding the percentage of missing values out of total cells 
total_cells = np.product(df.shape)  #finding rows*columns
total_missing = missing.sum()  #from missing finding sum (gives total of null values in all columns)
# percent of data that is missing
percent_missing = (total_missing/total_cells) * 100
print(percent_missing)

4.156490728506051


In [17]:
df.describe()
#gives values only for the columns with numeric datatype

start_lat     start_lng       end_lat       end_lng
count  5.723532e+06  5.723532e+06  5.718816e+06  5.718816e+06
mean   4.190065e+01 -8.764629e+01  4.190091e+01 -8.764650e+01
std    4.649557e-02  2.974885e-02  4.657768e-02  2.935251e-02
min    4.164000e+01 -8.784000e+01  4.139000e+01 -8.897000e+01
25%    4.188132e+01 -8.766014e+01  4.188132e+01 -8.766028e+01
50%    4.189862e+01 -8.764194e+01  4.189897e+01 -8.764288e+01
75%    4.192889e+01 -8.762775e+01  4.192914e+01 -8.762775e+01
max    4.563503e+01 -7.379648e+01  4.216812e+01 -8.749000e+01

### Checking for duplicate values 

In [18]:
#ride_id should have unique values 
df["ride_id"].nunique()  
#there are 5723532 unique values in ride_id column total rows=5723532

5723532

In [19]:
# checking the number of unique values per column
print("Unique values per column")
for col in df.columns:
    print(f"{col}: {df[col].nunique()}")

Unique values per column
ride_id: 5723532
rideable_type: 3
started_at: 4793191
ended_at: 4787114
start_station_name: 861
start_station_id: 852
end_station_name: 860
end_station_id: 852
start_lat: 433057
start_lng: 411835
end_lat: 434703
end_lng: 396215
member_casual: 2


- "ride_id", presumably the intended primary key of this table, clearly has  no duplicates since the number of unique values is equal to the data frame's size(records)
- "rideable_type" and "member_casual" both have a limited number of unique values
- There are more station_name than station id, which implies that many station IDS are associated with multiple station name (one id for multtiple station name as id :852< name:861)

In [20]:
stations_df = pd.read_csv("Divvy_Bicycle_Stations.csv", usecols = ['ID', 'Station Name', 'Latitude', 'Longitude'], index_col = 'Station Name')
stations_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 846 entries, Troy St & Jackson Blvd to McClurg Ct & Ohio St
Data columns (total 3 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   ID         846 non-null    int64  
 1   Latitude   846 non-null    float64
 2   Longitude  846 non-null    float64
dtypes: float64(2), int64(1)
memory usage: 26.4+ KB


In [21]:
stations_df.head()

ID   Latitude  Longitude
Station Name                                                         
Troy St & Jackson Blvd      1594046383808271024  41.877505 -87.704850
Western Ave & Fillmore St                   644  41.868563 -87.686234
Indiana Ave & Roosevelt Rd                  255  41.867888 -87.623041
Prospect Sq & 91st St                       684  41.728234 -87.667524
Fairbanks Ct & Grand Ave                     24  41.891847 -87.620580

In [22]:
def get_inconsistent_ids_series():
    """Get a pandas series of stations sharing the same ID and their value counts"""
    start_ids_sr = df.groupby(['start_station_id','start_station_name']).size().rename_axis(index={'start_station_id': 'id', 'start_station_name': 'station_name'})
    end_ids_sr = df.groupby(['end_station_id','end_station_name']).size().rename_axis(index={'end_station_id': 'id', 'end_station_name': 'station_name'})

    vc1 = start_ids_sr.reset_index()['id'].value_counts()
    vc2 = end_ids_sr.reset_index()['id'].value_counts()

    start_ids_with_many_names = vc1[vc1 > 1].index
    end_ids_with_many_names = vc2[vc2 > 1].index

    start_mask = start_ids_sr.index.get_level_values(0).isin(start_ids_with_many_names)
    start_ids_sr[start_mask]
    end_mask = end_ids_sr.index.get_level_values(0).isin(end_ids_with_many_names)
    end_ids_sr[end_mask]

    return pd.concat([start_ids_sr[start_mask], end_ids_sr[end_mask]], axis = 0)

inconsistent_ids_series = get_inconsistent_ids_series()
print("Number of station IDs with more than one name associated with it:", len(inconsistent_ids_series))
inconsistent_ids_series.head(10)

Number of station IDs with more than one name associated with it: 40


id      station_name                      
13099   Halsted St & 18th St                   1752
        Halsted St & 18th St (Temp)            1830
13300   DuSable Lake Shore Dr & Monroe St     23459
        Lake Shore Dr & Monroe St             22133
351     351                                       2
        Mulligan Ave & Wellington Ave           317
444     N Shore Channel Trail & Argyle Ave       11
        N Shore Channel Trail & Argyle St        32
LF-005  DuSable Lake Shore Dr & North Blvd    25767
        Lake Shore Dr & North Blvd            21996
dtype: int64

In [23]:
start_ids_sr = df.groupby(['start_station_id','start_station_name']).size().rename_axis(index={'start_station_id': 'id', 'start_station_name': 'station_name'})
start_ids_sr 

id                                     station_name                         
13001                                  Michigan Ave & Washington St             25878
13006                                  LaSalle St & Washington St               10623
13008                                  Millennium Park                          42287
13011                                  Canal St & Adams St                      18445
13016                                  St. Clair St & Erie St                   29948
                                                                                ...  
Throop/Hastings Mobile Station         Throop/Hastings Mobile Station               2
WL-008                                 Clinton St & Roosevelt Rd                 8405
WL-011                                 Jefferson St & Monroe St                  9131
WL-012                                 Clinton St & Washington Blvd             22421
Wilton Ave & Diversey Pkwy - Charging  Wilton Ave & Diversey Pk

In [24]:
vc1 = start_ids_sr.reset_index()['id'].value_counts()
vc1   #taking only the id column 
#finding count for each unique values in id

TA1306000029                             3
13300                                    2
351                                      2
LF-005                                   2
13099                                    2
                                        ..
20233                                    1
20234                                    1
20235                                    1
20236                                    1
Wilton Ave & Diversey Pkwy - Charging    1
Name: id, Length: 852, dtype: int64

In [25]:
vc1[vc1 > 1]  #id where vc1 is greater than 1

TA1306000029    3
13300           2
351             2
LF-005          2
13099           2
TA1305000039    2
TA1309000039    2
TA1309000049    2
444             2
TA1307000041    2
Name: id, dtype: int64

In [26]:
start_ids_with_many_names = vc1[vc1 > 1].index  #taking only the ids
start_ids_with_many_names

Index(['TA1306000029', '13300', '351', 'LF-005', '13099', 'TA1305000039',
       'TA1309000039', 'TA1309000049', '444', 'TA1307000041'],
      dtype='object')

In [27]:
start_ids_sr.index()  #as the index values are not unique and has duplicate we get error

TypeError: 'MultiIndex' object is not callable

In [28]:
start_ids_sr.index.get_level_values(0)

Index(['13001', '13006', '13008', '13011', '13016', '13017', '13021', '13022',
       '13028', '13029',
       ...
       'TA1309000061', 'TA1309000063', 'TA1309000064', 'TA1309000066',
       'TA1309000067', 'Throop/Hastings Mobile Station', 'WL-008', 'WL-011',
       'WL-012', 'Wilton Ave & Diversey Pkwy - Charging'],
      dtype='object', name='id', length=863)

In [29]:
start_ids_sr.index.get_level_values(0).isin(start_ids_with_many_names)

array([False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
        True,  True, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False,  True,  True, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False,

In [30]:
start_mask = start_ids_sr.index.get_level_values(0).isin(start_ids_with_many_names)
start_ids_sr[start_mask]

id            station_name                          
13099         Halsted St & 18th St                       1752
              Halsted St & 18th St (Temp)                1830
13300         DuSable Lake Shore Dr & Monroe St         23459
              Lake Shore Dr & Monroe St                 22133
351           351                                           2
              Mulligan Ave & Wellington Ave               317
444           N Shore Channel Trail & Argyle Ave           11
              N Shore Channel Trail & Argyle St            32
LF-005        DuSable Lake Shore Dr & North Blvd        25767
              Lake Shore Dr & North Blvd                21996
TA1305000039  Elston Ave & Cortland St                   7543
              Marshfield Ave & Cortland St               1495
TA1306000029  DuSable Lake Shore Dr & Ohio St           10476
              Lake Shore Dr & Ohio St                    9534
              McClurg Ct & Ohio St                       1490
TA1307000041  DuS

It looks like IDs can associate with several station names due to the time series nature of our data set and the evolution of station names and locations from events like relocations, renovations, and renamings. We'll move forward assuming that these instances are valid since they point to the same station.

## Transforming and cleaning the data

In [31]:
df_mem_usage = df.memory_usage(deep = True)/1_000_000
df_mem_usage # give the size of the columns in df in MB

Index                   0.000128
ride_id               417.817836
rideable_type         396.788571
started_at             45.788256
ended_at               45.788256
start_station_name    425.907421
start_station_id      349.744385
end_station_name      423.504240
end_station_id        348.071621
start_lat              45.788256
start_lng              45.788256
end_lat                45.788256
end_lng                45.788256
member_casual         360.582516
dtype: float64

In [32]:
def cols_mem_usage(*args: str) -> None:
    """Print dataframe column's memory usage in megabytes."""
    df_mem_usage = df.memory_usage(deep = True)
    for col_name in list(args):
        print(f"{col_name} total memory usage: {df_mem_usage[col_name]/1_000_000}MB")

#converting rideable_type and member_casual to category datatype
#The category data type in pandas is a hybrid data type. 
#It looks and behaves like a string in many instances but internally is represented by an array of integers. 
#This allows the data to be sorted in a custom order and to more efficiently store the data
print("Before:")
cols_mem_usage('rideable_type','member_casual') #calling the function 

# let's convert these low unique count columns into categories (categories is a effiecient datatype in terms of memory)
df['rideable_type'] = df['rideable_type'].astype('category')
df['member_casual'] = df['member_casual'].astype('category')

print("After:")
cols_mem_usage('rideable_type','member_casual')  #calling the function

Before:
rideable_type total memory usage: 396.788571MB
member_casual total memory usage: 360.582516MB
After:
rideable_type total memory usage: 5.723847MB
member_casual total memory usage: 5.723766MB


In [33]:
#function to get no of dupliate value in columns
def get_duplicates(field: str):  #input field in  str datatype
    vc = df[field].value_counts()
    duplicate_index = vc[vc > 1].index.to_list()
    n_duplicates = len(duplicate_index)   #no of duplicates is len of the list
    print(f"Number of duplicates for {field}: {n_duplicates}")
    if n_duplicates > 0:
        print("Sample rows:")
        mask = df[field].isin(duplicate_index)  # isin is like a in operator in python
        return df[mask].sort_values(by = field).head(4) #printing four rows which has duplicate records
    return

get_duplicates('ride_id')

Number of duplicates for ride_id: 0


In [34]:
vc = df['ride_id'].value_counts()  
vc  #gives a series object as ouput
#index            value

6C992BD37A98A63F    1
C959654606C3E9CF    1
1743B92404FCEBCF    1
D04E5B8D57701D84    1
7130D5B74EBAE98A    1
                   ..
7E64F0F87A1ECFD2    1
B33F2286F8062A00    1
4FE09D3068AC20F7    1
BADD19DC44B4B99C    1
5AEC5F39AF711895    1
Name: ride_id, Length: 5723532, dtype: int64

In [35]:
#getting only those records index where value>1 and converting to list
duplicate_index = vc[vc > 1].index.to_list() 
duplicate_index

[]

##  Finding ride _duration in negative 

In [37]:
negative_ride = (df['ended_at'] - df['started_at']) < pd.Timedelta("0 days")
print("Number of rows with negative ride durations:", len(df.loc[negative_ride]))

Number of rows with negative ride durations: 145


In [38]:
pd.Timedelta("0 days")   # just 0 days 

Timedelta('0 days 00:00:00')

In [39]:
(df['ended_at'] - df['started_at'])  #output is in timedelta format 

0         0 days 00:31:19
1         0 days 01:04:18
2         3 days 22:57:39
3         0 days 00:25:06
4         0 days 01:31:17
                ...      
5723527   0 days 00:08:29
5723528   0 days 00:46:12
5723529   0 days 00:13:28
5723530   0 days 00:05:40
5723531   0 days 00:09:59
Length: 5723532, dtype: timedelta64[ns]

In [40]:
(df['ended_at'] - df['started_at'])< pd.Timedelta("0 days") #check and give rows where days is less than 0 days 

0          False
1          False
2          False
3          False
4          False
           ...  
5723527    False
5723528    False
5723529    False
5723530    False
5723531    False
Length: 5723532, dtype: bool

In [41]:
df["Ride_duration"]=(df['ended_at'] - df['started_at']) #output is in timedelta format

In [42]:
df.head(1)  #Ride_duration column is added 

ride_id rideable_type          started_at            ended_at  \
0  6C992BD37A98A63F  classic_bike 2021-04-12 18:25:36 2021-04-12 18:56:55   

      start_station_name start_station_id              end_station_name  \
0  State St & Pearson St     TA1307000061  Southport Ave & Waveland Ave   

  end_station_id  start_lat  start_lng   end_lat   end_lng member_casual  \
0          13235  41.897448 -87.628722  41.94815 -87.66394        member   

    Ride_duration  
0 0 days 00:31:19

In [43]:
negative_ride= df["Ride_duration"]< pd.Timedelta("0 days")
negative_ride

0          False
1          False
2          False
3          False
4          False
           ...  
5723527    False
5723528    False
5723529    False
5723530    False
5723531    False
Name: Ride_duration, Length: 5723532, dtype: bool

In [44]:
df[negative_ride]  #df.loc[] and df[]  are almost same 

ride_id  rideable_type          started_at  \
22361    BC53ECCBC76278FD   classic_bike 2021-04-07 16:11:33   
31844    209C097828F9CD43  electric_bike 2021-04-27 17:13:44   
292566   6E81034B446FC2FD  electric_bike 2021-04-23 09:43:39   
292678   318DD838369AEA61   classic_bike 2021-04-30 10:56:32   
293033   8ADD13BD8F6A7567   classic_bike 2021-04-17 12:43:36   
...                   ...            ...                 ...   
4969725  5AA2BC364BC7A569  electric_bike 2021-11-07 01:59:53   
4971172  F4E4485BFB33D916  electric_bike 2021-11-07 01:57:53   
4972542  B506DCD44974C575  electric_bike 2021-11-07 01:53:34   
5623702  2D97E3C98E165D80   classic_bike 2022-03-05 11:00:57   
5626879  7407049C5D89A13D  electric_bike 2022-03-05 11:38:04   

                   ended_at                      start_station_name  \
22361   2021-04-07 16:11:26                 Ashland Ave & Grand Ave   
31844   2021-04-27 17:11:32                                     NaN   
292566  2021-04-23 09:43:29                   Dayton St & North Ave   
292678  2021-04-30 10:56:30                   Dayton St & North Ave   
293033  2021-04-17 12:43:27                   Dayton St & North Ave   
...                     ...                                     ...   
4969725 2021-11-07 01:09:02                                     NaN   
4971172 2021-11-07 01:27:02                                     NaN   
4972542 2021-11-07 01:00:42           Milwaukee Ave & Fullerton Ave   
5623702 2022-03-05 10:55:01  DuSable Lake Shore Dr & Wellington Ave   
5626879 2022-03-05 11:37:57          Sheffield Ave & Wellington Ave   

        start_station_id                        end_station_name  \
22361              13434                 Ashland Ave & Grand Ave   
31844                NaN                                     NaN   
292566             13058                   Dayton St & North Ave   
292678             13058                   Dayton St & North Ave   
293033             13058                   Dayton St & North Ave   
...                  ...                                     ...   
4969725              NaN                                     NaN   
4971172              NaN                                     NaN   
4972542              428                                     NaN   
5623702     TA1307000041  DuSable Lake Shore Dr & Wellington Ave   
5626879     TA1307000052          Sheffield Ave & Wellington Ave   

        end_station_id  start_lat  start_lng    end_lat    end_lng  \
22361            13434  41.891072 -87.666611  41.891072 -87.666611   
31844              NaN  41.910000 -87.640000  41.910000 -87.640000   
292566           13058  41.910638 -87.649372  41.910650 -87.649386   
292678           13058  41.910578 -87.649422  41.910578 -87.649422   
293033           13058  41.910578 -87.649422  41.910578 -87.649422   
...                ...        ...        ...        ...        ...   
4969725            NaN  41.920000 -87.690000  41.900000 -87.700000   
4971172            NaN  41.910000 -87.720000  41.900000 -87.770000   
4972542            NaN  41.920000 -87.700000  41.910000 -87.710000   
5623702   TA1307000041  41.936688 -87.636829  41.936688 -87.636829   
5626879   TA1307000052  41.936313 -87.652522  41.936253 -87.652662   

        member_casual     Ride_duration  
22361          member -1 days +23:59:53  
31844          member -1 days +23:57:48  
292566         member -1 days +23:59:50  
292678         member -1 days +23:59:58  
293033         member -1 days +23:59:51  
...               ...               ...  
4969725        casual -1 days +23:09:09  
4971172        casual -1 days +23:29:09  
4972542        casual -1 days +23:07:08  
5623702        casual -1 days +23:54:04  
5626879        casual -1 days +23:59:53  

[145 rows x 14 columns]

In [45]:
print("Number of rows with negative ride durations:", len(df.loc[negative_ride]))
# number of negative values  (look at ride_duration column value are in negative)

Number of rows with negative ride durations: 145


In [46]:
df=df[~negative_ride]  #only taking columns that have ride_duration positive values 

In [47]:
negative_ride= df["Ride_duration"]< pd.Timedelta("0 days")
negative_ride
print("Number of rows with negative ride durations:", len(df.loc[negative_ride]))
#all the rows with negative values is deleted 

Number of rows with negative ride durations: 0


# MISSING VALUES 